core/app/models/spree/address.rb

```diff
@@ -8,15 +8,22 @@ module Spree
   class Address < Spree::Base
     extend ActiveModel::ForbiddenAttributesProtection
 
+    mattr_accessor :state_validator_class,
+                   :state_normalizer_class
+    self.state_validator_class = Spree::Address::StateValidator
+    self.state_normalizer_class = Spree::Address::StateNormalizer
+
     belongs_to :country, class_name: "Spree::Country"
     belongs_to :state, class_name: "Spree::State"
 
     validates :firstname, :address1, :city, :country_id, presence: true
     validates :zipcode, presence: true, if: :require_zipcode?
     validates :phone, presence: true, if: :require_phone?
 
-    validate :state_validate
-    validate :validate_state_matches_country
+    validate do
+      self.class.state_normalizer_class.new(self).perform!
```

Comment:
Normalization is also happening into the validation, right? Why it is needed to perform it here as well?

Reply:
The original idea was to move it here from the validator service object so they can be completely decoupled. Just forgot to remove it from there (see also the comment bellow related to this one)